### PR TITLE
Edit regional partner

### PR DESCRIPTION
**Pull request by sureshc**
Workshop Admins can search, view, create, and edit (but not delete) Regional Partners.

search regional partners

edit regional partner

### CODE REVIEW

**Comment by sureshc** [comment]
I'm still working on tests for the controller update and create actions.

File: apps/src/code-studio/pd/workshop_dashboard/workshop_index.jsx [hunk 1]
```diff
@@ -91,6 +95,13 @@ const WorkshopIndex = React.createClass({
             >
               User Management
             </Button>}
+          {this.permission.isWorkshopAdmin &&
+            <Button
+              href={this.context.router.createHref("../../regional_partners")}
+              onClick={this.handleRegionalPartnersClick}
+            >
+              Regional Partners
```
Comment (aoby) [inline apps/src/code-studio/pd/workshop_dashboard/workshop_index.jsx]: nit: ideally button text should have a verb. Suggestion "Manage Regional Partners"

More generally do we want to clutter this page with so many buttons? Maybe we should have a single admin page with multiple actions, and one button for that?

File: dashboard/app/controllers/regional_partners_controller.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
+class RegionalPartnersController < ApplicationController
+  authorize_resource class: :pd_regional_partner_management
```
Comment (aoby) [inline dashboard/app/controllers/regional_partners_controller.rb]: This one actually has a model, `RegionalPartner`. Why not use that?

File: dashboard/app/controllers/regional_partners_controller.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  before_action :set_regional_partner, only: [:show, :edit, :update]
```
Comment (aoby) [inline dashboard/app/controllers/regional_partners_controller.rb]: Instead of doing this manually, leverage CanCan's `load_resource`. Even better, combine this with the previous line and use [load_and_authorize_resource](https://github.com/CanCanCommunity/cancancan/wiki/authorizing-controller-actions)

File: dashboard/app/controllers/regional_partners_controller.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    @regional_partners = RegionalPartner.all unless @regional_partners.try(:any?)
+  end
+
+  # GET /regional_partners/:id
+  def show
+  end
+
+  # GET /regional_partners/new
+  def new
+    @regional_partner = RegionalPartner.new
```
Comment (aoby) [inline dashboard/app/controllers/regional_partners_controller.rb]: note `load_resource` or `load_and_authorize_resource` (suggested above) will handle this for you

File: dashboard/app/controllers/regional_partners_controller.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  end
+
+  # POST /regional_partners
+  def create
+    @regional_partner = RegionalPartner.new(regional_partner_params)
```
Comment (aoby) [inline dashboard/app/controllers/regional_partners_controller.rb]: This is also not needed if using CanCan properly (see above)

File: dashboard/app/controllers/regional_partners_controller.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+      render action: 'edit'
+    end
+  end
+
+  private
+
+  # Use callbacks to share common setup or constraints between actions.
+  def set_regional_partner
+    @regional_partner = RegionalPartner.find(params[:id])
+  end
```
Comment (aoby) [inline dashboard/app/controllers/regional_partners_controller.rb]: remove. Use standard CanCan

File: dashboard/app/models/ability.rb [hunk 7]
```diff
@@ -128,6 +129,7 @@ def initialize(user)
         can :manage, :pd_teacher_attendance_report
         can :manage, Pd::TeacherApplication
         can :manage, :pd_workshop_user_management
+        can :manage, :pd_regional_partner_management
```
Comment (aoby) [inline dashboard/app/models/ability.rb]: (as mentioned above) Why not use the actual class? 

```ruby
can :manage, RegionalPartner
```

File: dashboard/app/views/regional_partners/_form.html.haml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  .field
+    = f.label :zip_code
+    = f.text_field :zip_code
+  .field
+    = f.label :phone_number
+    = f.text_field :phone_number
+  .field
+    = f.label :notes
+    = f.text_area :notes
+  %button.btn.btn-primary{type: 'submit', style: 'margin: 0'} Save
```
Comment (aoby) [inline dashboard/app/views/regional_partners/_form.html.haml:41]: What about contact? We can't set the contact?

What about other associated program managers?

File: dashboard/db/migrate/20170927162458_add_details_to_regional_partner.rb [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+    add_column :regional_partners, :attention, :string
+    add_column :regional_partners, :street, :string
+    add_column :regional_partners, :apartment_or_suite, :string
+    add_column :regional_partners, :city, :string
+    add_column :regional_partners, :state, :string
+    add_column :regional_partners, :zip_code, :string
+    add_column :regional_partners, :phone_number, :string
+    add_column :regional_partners, :notes, :text
+  end
+end
```
Comment (aoby) [inline dashboard/db/migrate/20170927162458_add_details_to_regional_partner.rb]: The migration should be separated into its own PR.

File: dashboard/db/migrate/20170927170055_add_timestamps_to_regional_partner.rb [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+  def remove_timestamps_for(model)
+    remove_column model.table_name, :created_at
+    remove_column model.table_name, :updated_at
+  end
+
+  def now
+    @_now ||= Time.zone.now
+  end
+end
```
Comment (aoby) [inline dashboard/db/migrate/20170927170055_add_timestamps_to_regional_partner.rb]: This too. Migrations should be isolated in their own PRs for cleaner tracking and rollback.

Comment (aoby) [reply dashboard/db/migrate/20170927170055_add_timestamps_to_regional_partner.rb]: These 2 migrations could be combined (either in one migration, or 2 migrations in a single PR) but in any case they should be separate from the model changes.

**Comment by tanyaparker** [comment]
[Make Regional Partner entity admin-manageable](https://trello.com/c/lUX2RBZB/156-make-regional-partner-entity-admin-manageable)

**Comment by sureshc** [comment]
@aoby PTAL

File: apps/src/code-studio/pd/workshop_dashboard/workshop_index.jsx (hunk 1 above)
Comment (sureshc) [reply apps/src/code-studio/pd/workshop_dashboard/workshop_index.jsx]: @tanyaparker Any thoughts on where links to PD admin features should live such as assigning courses to facilitators and editing regional partners?

File: dashboard/app/views/regional_partners/_form.html.haml (hunk 8 above)
Comment (sureshc) [reply dashboard/app/views/regional_partners/_form.html.haml:41]: I excluded edit of contacts since that feature is being deprecated in favor of assigning program managers in #18043

File: dashboard/app/models/regional_partner.rb [hunk 11]
```diff
@@ -33,6 +34,57 @@ class RegionalPartner < ActiveRecord::Base
...
+        errors.add(:zip_code, "#{DOES_NOT_MATCH_ADDRESS} #{found.first.postal_code}?")
+      end
+      if found.first.state_code != state
+        errors.add(:state, "#{DOES_NOT_MATCH_ADDRESS} #{found.first.state_code}?")
+      end
+      unless found.first.street_number
+        errors.add(:street, INVALID_STREET_ADDRESS)
+      end
+    end
+  end
```
Comment (aoby) [inline dashboard/app/models/regional_partner.rb]: Are we sure we want to fully validate the address here? Occasionally real addresses aren't found correctly by the geocoder, and especially don't return with a street number. In the workshop material order [form](https://github.com/code-dot-org/code-dot-org/blob/e38b71109b55f218afa5c8548146354e6c56d0c1/dashboard/app/views/pd/workshop_material_orders/new.html.haml#L24) we added a [manual override](https://github.com/code-dot-org/code-dot-org/blob/735a3dce8f0d77dd02b5d63d6eed7386f9692449/dashboard/app/models/pd/workshop_material_order.rb#L66) "I verify that this address is correct, and I want to use it to ship my order" checkbox so teachers with "invalid" (according to the geocoder) street addresses could still order supplies. 

Do we care in this context if it's a PO box? Do we care to have such strict validation? If so we should at least provide an override.

Check with @tanyaparker but I don't think we need any full address validation here, beyond basic formatting.

File: dashboard/test/controllers/regional_partners_controller_test.rb [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    @regional_partner = create :regional_partner
+    @workshop_admin = create :workshop_admin
+  end
+
+  def self.test_workshop_admin_only(method, action, params = nil)
+    %i(student teacher facilitator workshop_organizer).each do |user_type|
+      test_user_gets_response_for action, user: user_type, method: method, params: params, response: :forbidden
+    end
+    test_user_gets_response_for action, user: -> {@workshop_admin}, method: method, params: params, response: :success
+  end
```
Comment (aoby) [inline dashboard/test/controllers/regional_partners_controller_test.rb:15]: nice :)

File: apps/src/code-studio/pd/workshop_dashboard/workshop_index.jsx (hunk 1 above)
Comment (tanyaparker) [reply apps/src/code-studio/pd/workshop_dashboard/workshop_index.jsx]: Here ya go! https://docs.google.com/document/d/1iaMcfalaH01Mw45Aze4iFqXh7ujXgBwhb0ybLrIunlg/edit

Comment (sureshc) [reply apps/src/code-studio/pd/workshop_dashboard/workshop_index.jsx]: I've removed the Regional Partners button on the Workshop Dashboard and will implement the new Workshop Admin tools page in a separate PR.

Comment (tanyaparker) [reply apps/src/code-studio/pd/workshop_dashboard/workshop_index.jsx]: 👍

**Comment by sureshc** [comment]
@tanyaparker @aoby PTAL

I replace address verification with validations on zip code and state.
I removed the "Regional Partner" button from the Workshop Dashboard and will implement the new Workshop Admin Tools page in a separate PR.
